File: core/sarva/agent/events.py
"""sarva.agent.events — the agent loop's state machine and event vocabulary.

Every skin (CLI, server, desktop) consumes only `AgentEvent`s — none of them
reach into loop internals. The state machine is explicit data (`LEGAL`) so
the loop's control flow stays inspectable rather than implicit in code paths.
"""

from __future__ import annotations

import time
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from sarva.agent.budget import Spend
from sarva.multimodal.content import Message, ToolCallBlock, ToolResultBlock
from sarva.providers.base import ProviderEvent


class AgentState(StrEnum):
    INIT = "init"
    CALLING_MODEL = "calling_model"
    RUNNING_TOOLS = "running_tools"
    AWAITING_CONFIRMATION = "awaiting_confirmation"
    DONE = "done"
    FAILED = "failed"
    INTERRUPTED = "interrupted"
    BUDGET_EXCEEDED = "budget_exceeded"


TERMINAL = {
    AgentState.DONE,
    AgentState.FAILED,
    AgentState.INTERRUPTED,
    AgentState.BUDGET_EXCEEDED,
}

LEGAL: dict[AgentState, set[AgentState]] = {
    AgentState.INIT: {AgentState.CALLING_MODEL},
    AgentState.CALLING_MODEL: {
        AgentState.RUNNING_TOOLS,
        AgentState.DONE,
        AgentState.FAILED,
        AgentState.INTERRUPTED,
        AgentState.BUDGET_EXCEEDED,
    },
    AgentState.RUNNING_TOOLS: {
        AgentState.AWAITING_CONFIRMATION,
        AgentState.CALLING_MODEL,
        AgentState.FAILED,
        AgentState.INTERRUPTED,
        AgentState.BUDGET_EXCEEDED,
    },
    AgentState.AWAITING_CONFIRMATION: {
        AgentState.RUNNING_TOOLS,
        AgentState.INTERRUPTED,
    },
}


class _AEvent(BaseModel):
    model_config = {"frozen": True}
    ts: float = Field(default_factory=time.time)


class StateChangedEvent(_AEvent):
    type: Literal["state_changed"] = "state_changed"
    state: AgentState
    detail: str | None = None


class ModelStreamEvent(_AEvent):
    """Provider deltas passed through for live rendering."""

    type: Literal["model_stream"] = "model_stream"
    event: ProviderEvent


class ToolStartedEvent(_AEvent):
    type: Literal["tool_started"] = "tool_started"
    call: ToolCallBlock


class ToolFinishedEvent(_AEvent):
    type: Literal["tool_finished"] = "tool_finished"
    result: ToolResultBlock
    seconds: float


class NeedsConfirmationEvent(_AEvent):
    type: Literal["needs_confirmation"] = "needs_confirmation"
    call: ToolCallBlock


class RunDoneEvent(_AEvent):
    type: Literal["run_done"] = "run_done"
    state: AgentState
    final_message: Message | None
    spend: Spend


AgentEvent = Annotated[
    (
        StateChangedEvent
        | ModelStreamEvent
        | ToolStartedEvent
        | ToolFinishedEvent
        | NeedsConfirmationEvent
        | RunDoneEvent
    ),
    Field(discriminator="type"),
]


class AgentResult(BaseModel):
    model_config = {"frozen": True}
    state: AgentState
    final_message: Message | None
    spend: Spend
    run_dir: str

File: apps/desktop/src/events.ts
/**
 * Typed mirror of sarva.agent.events.AgentEvent (core/sarva/agent/events.py).
 *
 * Kept minimal and local to this app rather than importing `sdks/typescript/`
 * directly — the SDK package now exists (it didn't when this comment
 * originally said "once a second consumer needs it"), but this app isn't
 * wired to depend on it yet; a real, still-open follow-up, not silently
 * assumed done. Field names match the Python Pydantic models' JSON output
 * exactly (snake_case), since that's the wire format over /ws/chat.
 *
 * A real drift bug found by a fresh-eyes sweep while investigating that
 * stale comment: this local copy's `run_done` variant was missing the
 * `spend` field entirely, even though `sarva.agent.events.RunDoneEvent`
 * (the real Pydantic model actually serialized over the wire) has
 * required it since `Spend` tracking shipped -- confirmed by reading
 * `RunDoneEvent` directly, not assumed from this file's own claim to be
 * an accurate mirror. `sdks/typescript/src/types.ts`'s own `AgentEvent`
 * already had it correctly; only this app's independent, never-reconciled
 * copy had silently drifted.
 *
 * A second instance of the identical drift class, found by comparing
 * this file against that same sibling SDK mirror directly rather than
 * just its own claim to be accurate: `ToolCall` here was a minimal stub
 * (`id`/`name`/`arguments`, no `type`) and `tool_finished`'s own
 * `result` field was `{ is_error: boolean }` only -- but the real
 * `sarva.multimodal.content.ToolCallBlock`/`ToolResultBlock` Pydantic
 * models (what actually crosses the wire) also carry `type`, and
 * `ToolResultBlock` additionally carries `tool_call_id` and `content`
 * (the tool's actual output) -- fields `sdks/typescript/src/types.ts`'s
 * own `ToolCallBlock`/`ToolResultBlock` already modeled completely.
 * `App.tsx` only ever reads `.name`/`.arguments`/`.is_error` today, so
 * this wasn't visibly broken -- but a future feature reading `event.
 * result.content` (e.g. actually showing what a tool returned, not just
 * ok/error) would have hit a field this "typed mirror" claimed didn't
 * exist, even though the server has always sent it.
 */
export interface Spend {
  model_calls: number;
  total_tokens: number;
  wall_seconds: number;
  cost_usd: number;
}

export type AgentState =
  | "init"
  | "calling_model"
  | "running_tools"
  | "awaiting_confirmation"
  | "done"
  | "failed"
  | "interrupted"
  | "budget_exceeded";

interface ToolCall {
  type: "tool_call";
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

/** Any single content block in a tool result (usually TextBlock, but
 * image/audio blocks are also possible) -- deliberately not fully
 * modeled here, matching `sdks/typescript/src/types.ts`'s own honest
 * "partial, not guessed" scoping for the same union. */
type ContentBlock = { type: string; [key: string]: unknown };

interface ToolResult {
  type: "tool_result";
  tool_call_id: string;
  content: ContentBlock[];
  is_error: boolean;
}

interface ProviderEvent {
  type: string;
  text?: string;
}

export type AgentEvent =
  | { type: "state_changed"; state: AgentState; detail?: string | null }
  | { type: "model_stream"; event: ProviderEvent }
  | { type: "tool_started"; call: ToolCall }
  | { type: "tool_finished"; result: ToolResult; seconds: number }
  | { type: "needs_confirmation"; call: ToolCall }
  | { type: "run_done"; state: AgentState; final_message: unknown; spend: Spend };
